include/odp/api/spec/queue.h

```diff
@@ -96,6 +96,32 @@ typedef enum odp_queue_op_mode_t {
 
 } odp_queue_op_mode_t;
 
+/**
+ * Blocking level
+ *
+ * Blocking level tells how operations behave when one of the threads using
+ * the same resource suspends. A blocking implementation may block other threads
+ * using the same resource for the entire time that one of the threads remain
+ * suspended. On the contrary, a lock-free implementation allows other threads
+ * to make progress while a thread is suspended. Blocking levels are listed from
+ * the weakest to the strongest guarantee of block freedom.
+ *
+ * Performance of non-blocking operations may be lower than blocking ones, but
+ * can be used safely e.g. between threads of real-time and non real-time parts
+ * of an application.
+ */
+typedef enum odp_blocking_level_t {
```

Comment:
Presumably you're talking about thread preemption here. Isn't that why we have dedicated cores? Why restrict this to just queues? `odp_packet_alloc()/free()` and other routines might be subject to the same considerations as well.

Reply:
If lockfreeness is not guaranteed, a thread interrupted during an enqueue/dequeue operation can block all other threads (spinning inside enq/deq operation). Dedicated thread may be also interrupted (or call system call) and control (slow path) threads typically share a core.

The type name is generic, so that we can extend this to other APIs later on. Queue is the most important one to fix for thread to thread communication.

Reply:
Why would the data plane thread get interrupted? The programming model for ODP has been that, the data plane threads are non preemptive.

Are you suggesting to change this programming model?

Reply:
This execution model might not hold for all types of ODP applications. E.g. control plane threads might run on non-dedicated cores and thus suffer interrupts and preemption. Even on dedicated and isolated cores, isolation could be imperfect (e.g. Linux 1Hz scheduling tick). Finally, HW is not perfect and deterministic and one limitation to scalability is (SW) dependencies between threads.